File: framework/src/main/java/org/synsystems/onlypass/framework/events/SharedPreferencesBackedEvent.java
package org.synsystems.onlypass.framework.events;

import android.content.SharedPreferences;
import android.support.annotation.NonNull;

import io.reactivex.Completable;
import io.reactivex.Single;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * An event that records occurrences in shared preferences.
 */
public class SharedPreferencesBackedEvent implements Event {
  private final SharedPreferences sharedPrefs;

  private final String key;

  /**
   * Constructs a new SharedPreferencesBackedEvent.
   *
   * @param sharedPrefs
   *     the shared preferences to store the record in
   * @param key
   *     the shared preferences key to use when storing the event
   */
  public SharedPreferencesBackedEvent(@NonNull final SharedPreferences sharedPrefs, @NonNull final String key) {
    this.sharedPrefs = checkNotNull(sharedPrefs);
    this.key = checkNotNull(key);
  }

  @NonNull
  @Override
  public Single<Boolean> hasOccurred() {
    return Single.fromCallable(() -> sharedPrefs.getBoolean(key, false));
  }

  @NonNull
  @Override
  public Completable recordOccurrence() {
    return Completable.fromRunnable(() -> sharedPrefs
        .edit()
        .putBoolean(key, true)
        .apply());
  }

  @NonNull
  @Override
  public Completable clearRecord() {
    return Completable.fromRunnable(() -> sharedPrefs
        .edit()
        .remove(key)
        .apply());
  }
}

File: framework/src/main/kotlin/org/synsystems/onlypass/framework/events/SharedPreferencesBackedEvent.kt
package org.synsystems.onlypass.framework.events

import android.content.SharedPreferences
import io.reactivex.Completable
import io.reactivex.Single

/**
 * An [Event] that is recorded in shared preferences.
 *
 * @property sharedPrefs the shared preferences to use for persistence
 * @property key the key to use when storing events in [sharedPrefs]
 */
class SharedPreferencesBackedEvent(
    private val sharedPrefs: SharedPreferences,
    private val key: String
) : Event {

  override fun hasOccurred() = Single.fromCallable { sharedPrefs.getBoolean(key, false) }

  override fun recordOccurrence() = Completable.fromRunnable {
    sharedPrefs
        .edit()
        .putBoolean(key, true)
        .apply()
  }

  override fun clearRecord() = Completable.fromRunnable {
    sharedPrefs
        .edit()
        .remove(key)
        .apply()
  }
}